config/config-api/src/com/thoughtworks/go/config/PipelineTemplateConfig.java

```diff
@@ -67,12 +69,62 @@ public CaseInsensitiveString name() {
         return name;
     }
 
+    public void validateTree(ValidationContext validationContext, CruiseConfig preprocessedConfig) {
+        validate(validationContext);
+        validateParamsAndFetchTasks(preprocessedConfig);
+    }
+
+    private void validateParamsAndFetchTasks(CruiseConfig preprocessedConfig) {
+        List<CaseInsensitiveString> pipelineNames = preprocessedConfig.pipelinesAssociatedWithTemplate(this.name());
+        for (CaseInsensitiveString pipelineName : pipelineNames) {
+            PipelineConfig pipelineConfig = preprocessedConfig.getPipelineConfigByName(pipelineName);
+            validateParams(pipelineConfig);
+            validateFetchTasks(preprocessedConfig, pipelineConfig);
+        }
+    }
+
+    private void validateFetchTasks(CruiseConfig preprocessedConfig, PipelineConfig pipelineConfig) {
+        PipelineConfigSaveValidationContext contextForStages = PipelineConfigSaveValidationContext.forChain(false, "", preprocessedConfig, pipelineConfig);
+        for (StageConfig stageConfig : pipelineConfig.getStages()) {
+            PipelineConfigSaveValidationContext contextForJobs = contextForStages.withParent(stageConfig);
+            for (JobConfig jobConfig : stageConfig.getJobs()) {
+                PipelineConfigSaveValidationContext contextForTasks = contextForJobs.withParent(jobConfig);
+                for (Task task : jobConfig.getTasks()) {
+                    if (task instanceof FetchTask) {
+                        task.validate(contextForTasks);
+                        this.errors().addAll(task.errors());
+                    }
+                }
+            }
+        }
+    }
+
+    private void validateParams(PipelineConfig pipelineConfig) {
+        for (ParamConfig paramConfig : this.referredParams()) {
```

Comment:
`this.referredParams()` does a deepclone of template-config. Since validateParams is being invoked for each dependent pipeline, the clone operation becomes a function that. You might want to pull it out of the pipeline loop.